nodejs/awsx/apigateway/apikey.ts

```diff
@@ -0,0 +1,149 @@
+     *  that has a combination of:
+     *  1 - an existing APIKey
+     *  2 - ApiKeyArgs for a new APIKey
+     */
+    apiKeys?: Array<aws.apigateway.ApiKey | aws.apigateway.ApiKeyArgs>;
+}
+
+export interface AssociatedAPIKeys {
+    readonly usagePlan: aws.apigateway.UsagePlan;
+    readonly keys: Key[];
+}
+
+export interface Key {
+    apikey: aws.apigateway.ApiKey;
+    usagePlanKey: aws.apigateway.UsagePlanKey;
+}
+
+/**
+ * Helper function that allows you to quickly create API Keys and associate them with an API.
+ *
+ * @param name The _unique_ name of the resource.
+ * @param args The arguments to use to populate this resource's properties.
+ */
+export function createAssociatedAPIKeys(name: string, args: APIKeyArgs): AssociatedAPIKeys {
+    const usagePlan = getUsagePlan(name, args);
+    const keys = getKeys(name, args, usagePlan);
+
+    return { usagePlan, keys };
+}
+
+/** @internal */
+function getUsagePlan(name: string, args: APIKeyArgs): aws.apigateway.UsagePlan {
+    if (args.usagePlan && pulumi.CustomResource.isInstance(args.usagePlan)) {
+        if (args.apis) {
+            throw new Error("cannot define both [args.apis] and an existing usagePlan [args.usagePlan]");
+        }
+        return args.usagePlan;
+    } else {
+        let usagePlanArgs: aws.apigateway.UsagePlanArgs | undefined = args.usagePlan;
+
+        if (args.apis) {
+            const stages = [];
+
+            for (const a of args.apis) {
+                stages.push({
+                    apiId: a.restAPI.id,
+                    stage: a.stage.stageName,
+                });
+            }
+
+            if (args.usagePlan) {
+                if (args.usagePlan.apiStages) {
+                    throw new Error("cannot define both [args.apis] and [args.usagePlan.apiStages]");
+                }
+
+                usagePlanArgs = {
+                    apiStages: stages,
+                    description: args.usagePlan.description,
+                    name: args.usagePlan.name,
+                    productCode: args.usagePlan.productCode,
+                    quotaSettings: args.usagePlan.quotaSettings,
+                    throttleSettings: args.usagePlan.throttleSettings,
+                };
+            } else {
+                usagePlanArgs = {
+                    apiStages: stages,
+                };
+            }
+        }
+        return new aws.apigateway.UsagePlan(name, usagePlanArgs);
+    }
+}
+
+/** @internal */
+function getKeys(name: string, args: APIKeyArgs, usagePlan: aws.apigateway.UsagePlan): Key[] {
+    const keys: Key[] = [];
+
+    if (args.apiKeys) {
+        for (let i = 0; i < args.apiKeys.length; i++) {
+            const currKey = args.apiKeys[i];
+            let apikey: aws.apigateway.ApiKey;
+
+            if (pulumi.CustomResource.isInstance(currKey)) {
+                apikey = currKey;
+            } else {
+                apikey = new aws.apigateway.ApiKey("${name}-${i}", currKey);
+            }
```

Comment:
tiny nit, and if you prefer this form feel free to keep it.  However, i find it easier to reason about 'const' variables.  And seeing blocks like this that are jsut used to initialize a single variable seems more complex than just having:

```ts
const apiKey = CustomResource.isInstance(currKey)
    ? currKey
    : new ...
```

Also note: our `?:` style is to wrap on the punctuation and align them indented on the next line.

--

If you prefer the if/else style though, that's fine.